### ISSUE TITLE
API (and internals) introducing 

### ISSUE THREAD

**Issue by askalt** [tarantool/go-tupleconv#1]
***raw parsers***

There are several raw parsers that parse strings and convert them into various types (`bool`, `uint64`, `int64`, and so on). These parsers can be used as utilities for writing custom parsers.

For example:
```Golang
func BoolParser(s string) (bool, error) 
```
Additionally, there is a function for combining parsers and applying them in a certain order:
```Golang
func MakeCompositeParser(parsers []Parser) Parser
```

***Converter***

`Converter` is the main type for constructing various converters. It is a abstract object, and its entire interface consists of one function:
```Golang
 Convert(val any, typ TypeName, isNullable bool) (any, error)
```
All this function does is find the conversion function in the conversion table based on the type name that needs to be executed for the conversion.  
The entire converter's functionality is configured using a `Config`, which looks like this:
```Golang
type Config struct {
	// DecSeparators are decimal separators.
	DecSeparators string
	// ThousandSeparators separate thousands.
	ThousandSeparators string
	// NullValue is a string interpreted as null.
	NullValue string

	// ConvertTable is a table, containing convert rules.
	ConvertTable map[TypeName]TypeConverter
}
```
The `ConvertTable` allows flexible customization. It is a mapping from type names to `TypeConverter`, which is declared as follows:
```Golang
type TypeConverter func(any, Config, bool) (any, error)
```
Exactly, the `TypeConverter` is a function that implements a certain type conversion. As you mentioned, its behavior can depend on the values from the `Config`. For example, a standard parser for converting strings to tarantool types will use the specified values from the configuration to perform the conversions. 
Third argument of  `TypeConverter` is `isNullable` flag, which is true, when this value can be `null`.

There are several built-in standard configs that cover the main use cases of the converter. These include:
- `TTDecoderConfig`, which converts from string to tarantool types
- `TTEncoderConfig`, which converts from tarantool types to string 

***Samples***

For example, this is how you can create the most basic converter from strings to tarantool types:
```Golang
decoder := tupleconv.NewConverter(tupleconv.TTDecoderConfig)
```

If custom settings are needed, you can do the following:
```Golang
decodeTable := tupleconv.TTDecodeTable
decodeTable["strange"] = func(s any, _ tupleconv.Config, isNullable bool) (any, error) {
	return s.(string) + "1", nil
}

decoder := tupleconv.NewConverter(tupleconv.Config{
	DecSeparators:        ".*",
	ThousandSeparators:   "@ '",
	NullValue:            "null",
	DecodeTable:          decodeTable,
})
```

Here is how the creation of an encoder looks like:
```Golang
encoder := tupleconv.NewConverter(tupleconv.TTEncoderConfig)
```

***TupleConverter***

`TupleConverter` is a helper type that allows converting spaces according to their format. To create it, you need to pass the parsed format of the space, as well as a specific converter.
```Golang
NewTupleConverter(conv *Converter, spaceFmt []FieldFmt) *TupleDecoder
```
It provides a function for converting a tuple according to the format:
```Golang
Decode(tuple any) ([]any, error)
```

For example, here's how you can create a parser for spaces (using converter with `TTDecoderConfig` as converter):
```Golang
tupleFormat := []any{
	map[any]any{"type": "unsigned"},
	map[any]any{"type": "integer", "is_nullable": true},
}
fmt, _ := tupleconv.ParseSpaceFmt(tupleFormat)

tupleDecoder := tupleconv.NewTupleConverter(
	tupleconv.NewConverter(tupleconv.TTDecoderConfig), fmt,
)

tup, _ := tupleDecoder.Convert([]string{"1", "null"})
assert.Equal(t, []any{uint64(1), nil}, tup)
```

In summary, the package provides flexible capabilities for creating type converters.

**Comment by oleg-jukovec** [comment]
Hello, I have a few questions.

> raw parsers

How does this relate to the rest of the idea and what interface has `Parser`?

> Convert(val any, typ TypeName, isNullable bool) (any, error)

What is `TypeName`?

> Converter is the main type for constructing various converters. It is a abstract object, and its entire interface consists of one function:

> The entire converter's functionality is configured using a Config.

Is `Converter` interface or a concrete implementation?

> // NullValue is a string interpreted as null.

We could work with `any` type, but a null value could be only a string?

> tupleFormat := []any{
>	map[any]any{"type": "unsigned"},
>	map[any]any{"type": "integer", "is_nullable": true},
> }
> fmt, _ := tupleconv.ParseSpaceFmt(tupleFormat)

This looks strange. What's the purpose of `map[any]any`, why not use a concrete type?

> Decode(tuple any) ([]any, error)

Why not `Decode(tuple []any) ([]any, error)` or `Decode(tuple any) (any, error)` ?

**Comment by oleg-jukovec** [comment]
I would suggest to start with something extremely simple and made it difficult if needed.

```Go
type Parser interface {
     Parse(src any) (any, error)
}
// + concrete implementations: BoolParser, IntegerParser, ...
//, ChainParser (sequential combination of parsers)
//, NullableParser (a wrapper around a regular parser + nullable)
//, some more wrappers/decorators
```
and
```Go
fmt := []Parser{BoolParser{}, IntegerParser{"@ '"}, NullableParser{DoubleParser{".*"}}}
default := AnyParser{}
decoder := NewConverter(fmt, default)
dst, err := decoder.Conver(src)
```
What functionality is missing here?

**Comment by askalt** [comment]
About raw_parsers.
They assist in using the package as described by you. The only difference is that initially, they were intended to be primitive parsers (with all the more complex logic implemented above), but I like your idea of simply calling them _Parsers_ and moving the arguments into them.

However, what about leaving the parsing format logic? It seems like a clear use case when there is something like this on the server-side:
```lua
function get_fmt(space):
   return box.space[space]:format()
end
```

Next on the client:
```golang
response, err := conn.Call("get_fmt", []any{"test"})
```
And we get exactly `[]any{map[any]any, ...}`.
Next, we can initialize the `TupleConverter` using it, and it parses it into the structure it needs, automatically constructing an array of parsers (`fmt` from your example).
How do you feel about this?

**Comment by oleg-jukovec** [comment]
> Next on the client:
> 
> ```go
> response, err := conn.Call("get_fmt", []any{"test"})
> ```
> 
> And we get exactly `[]any{map[any]any, ...}`.

You could use CallTyped and get a filled structure.

> Next, we can initialize the TupleConverter using it, and it parses it into the structure it needs, automatically constructing an array of parsers (fmt from your example).

I don't see a big difference here, you could use a helper code on the client side to construct the `[]Parsers` slice.

**Comment by askalt** [comment]
Agreed, but let's imagine about what this helper code would look like.

On the client-side, we need to handle any type, which means we should be able to provide a `Parser` for any type.
Hence, most-likely, this will be implemented through some kind of table `map[string]Parser`, and it will be a quite large table, essentially repeating from client to client (since the most common case is the standard decoding/encoding of tarantool types, the tables won't differ much).

So, why not introduce such logic in the library itself, creating some basic tables that can be configured in some way, so that client doesn't have to think about how to build these arrays of parsers in most cases?

**Comment by psergee** [comment]
> Next, we can initialize the TupleConverter using it, and it parses it into the structure it needs, automatically constructing an array of parsers (fmt from your example).

I'd like to avoid relying on specific data structure returned from any lua calls in library. If I ever want to create a tuple from my slices, I will need to generate very specific map structure to init converter. But it is much simpler to provide a slice of implemented serializers/deserializers. We may have a separate function which generates a slice of [de]serializers from `lua format style`.

**Comment by psergee** [comment]
> So, why not introduce such logic in the library itself, creating some basic tables that can be configured in some way, so that client doesn't have to think about how to build these arrays of parsers in most cases?

It is convenient to have such functionality in library. But still don't like the idea of relying on specific `any` map returned from `lua`. We can have a function that creates a [de]serializer for the specified type, which can be specified using `string` or `any`. So, if a user has space format in any form, the code will be something like:
```
mySerializers := SerializerInterface[]
for _, typeName := range myFormat {
    mySerializers.Push(NewSerializerFor(typeName))
}
```
So the library does not know something about tarantool space format structure.

**Comment by oleg-jukovec** [comment]
> So, why not introduce such logic in the library itself, creating some basic tables that can be configured in some way, so that client doesn't have to think about how to build these arrays of parsers in most cases?

It will not be too difficult to add such functions/types to the library.

```Go
fmt := GetStringsToTTFormat(someKindOfTheSpaceFormat) // returns []Parser
```

**Comment by psergee** [comment]
> The entire converter's functionality is configured using a Config, which looks like this:
  type Config struct {
	  // DecSeparators are decimal separators.
	  DecSeparators string
	  // ThousandSeparators separate thousands.
	  ThousandSeparators string
	  // NullValue is a string interpreted as null.
	  NullValue string
	  // ConvertTable is a table, containing convert rules.
	  ConvertTable map[TypeName]TypeConverter
  }

Let's avoid global configuration. It violates SRP. All customizations must be done in [de]serializers. And Converter implementation must depend only on interfaces.

**Comment by askalt** [comment]
Summarize:

- [de]serializers/converters is type that satisfy the following interface:
```golang
type Converter interface {
	Convert(src any) (any, error)
}
```
All necessary settings are stored inside a specific implementation like `IntConverter { ThousandSeparators }`.

- `TupleConverter`  initialized with converters list: `[]Converter`.
-  Having some space format, the user must create converters list manually, or using our helpers like `NewSerializerFor(bool)`, or more automatic `GetStringsToTTFormat`.

**Comment by psergee** [comment]
> type Converter interface {
	Convert(src any) (any, error)
}

Will we have two implementations for converting string -> int and int -> string? Or is it supposed to be one interface to convert forward and back?

**Comment by askalt** [comment]
> > type Converter interface {
> > Convert(src any) (any, error)
> > }
> 
> Will we have two implementations for converting string -> int and int -> string? Or is it supposed to be one interface to convert forward and back?

It seems that two implementations are more flexible, because:
- In some cases you don't want do reverse conversion 
- In some cases there is no suitable reverse conversion 

I suppose at least three basic implementations of `Converter`:
- `Deserializer` is a converter from string to any (i.e., parser)
- `Serializer` is a converter from any to string 
- `FuncConverter` is just a box for convenient `Converter` creation: it takes a function and creates Converter based on this

**Comment by psergee** [comment]
> It seems that two implementations are more flexible

Right.

> I suppose at least three basic implementations of Converter:
Deserializer is a converter from string to any (i.e., parser)

So, can deserializer use string as input an not any? The same is for output of deserializer.

**Comment by psergee** [comment]
What I mean... We can make our high level Converter type more flexible to avoid working with `any` for most of our scenarios. Something like:
```
type Deserializer[T any] interface {
	Convert(in T) any
}

type IntDeserializer struct {
}

func (IntDeserializer) Convert(string) any {
	return 42
}

type IntDeserializerEnhanced struct {
}

func (IntDeserializerEnhanced) Convert(any) any {
	return 53
}

type Converter[ConvType any] struct {
	converters []Deserializer[ConvType]
}
```
So, for CSV input I can create the Converter, which uses deserializers accepting only strings:
```
	var c Converter[string]
	c.converters = append(c.converters, IntDeserializer{})
	fmt.Println(c.converters[0].Convert("1"))
```
If I ever want something more euristic:
```
	var d Converter[any]
	d.converters = append(d.converters, IntDeserializerEnhanced{})
	fmt.Println(d.converters[0].Convert("1"))
```

**Comment by askalt** [comment]
> So, can deserializer use string as input an not any? The same is for output of deserializer. 

I thought to allow the Deserializer to accept any as input but still validate it:
```golang
func (des *Deserializer) Convert(src any) (any, error) {
	srcAsStr, ok := src.(string)
	if !ok {
		return nil, NewErrUnexpectedValue(src)
	}
	return des.parse(srcAsStr)
}
```

Of course, we can prohibit this in type level by using generics:
```golang
type Deserializer = Converter[string, any]
```
But I think it might lead to overly complex code, and the game isn't worth the candle. What do you think?

**Comment by psergee** [comment]
> I thought to allow the Deserializer to accept any as input but still validate it:

I'd like to avoid any for most cases. It is better to do not check type of any if it is possible.
